### PR TITLE
Kokkos_ENABLE_Cuda_Lambda default ON if CUDA >= 75

### PR DESCRIPTION
**Pull request by ibaned**
[#1101]

Spot checked that it does default to ON in a CUDA 8.0 build of Trilinos

### CODE REVIEW

**Comment by crtrott** [comment]
What does GEQ stand for?

**Comment by crtrott** [comment]
Also do we really want to turn this on by default, if that means people have to add no obvious flags to their CMAKE_CXX_FLAGS?

**Comment by ibaned** [comment]
> What does GEQ stand for?

"Greater than or EQual to"

> do we really want to turn this on by default, if that means people have to add no obvious flags to their CMAKE_CXX_FLAGS?

Yes, I think so. People trying to use lambdas right now already have to remember to do both those things, this brings it down to one thing to remember. Also, we should definitely work on getting `-expt-extended-lambda` into Trilinos flags by getting Kokkos logic in to their build system.

I guess the only case this causes trouble is people using `KOKKOS_LAMBDA`, only passing them into policies with OpenMP or Serial space, and still compiling with CUDA, and also not adding the `-expt-extended-lambda` flag. Thats a pretty strange situation, and we can fix even that by checking for the flag in `CMAKE_CXX_FLAGS`.

**Comment by crtrott** [comment]
If Trilinos wants this fine. I think it is a mistake and Trilinos should protect its LAMBDA usage, instead of forcing everybody to add experimental compiler flags to their build in non-obvious ways .....

**Comment by mhoemmen** [comment]
@crtrott wrote:
> If Trilinos wants this fine. I think it is a mistake and Trilinos should protect its LAMBDA usage, instead of forcing everybody to add experimental compiler flags to their build in non-obvious ways .....

I feel the need to push back on this a bit.  First, lambdas worked fine without that build option before Kokkos changed its macros.  I'm not convinced that the build option is necessary, and I need to do more experiments to see if that is the case.  Second, while Trilinos might build and pass tests with CUDA, that doesn't mean its solvers are all ready for production use with CUDA.  I would rather write to the language that we will likely have by the time we need production-ready CUDA solvers, than do elaborate work-arounds to support an unfunded use case.

**Comment by crtrott** [comment]
Mark I am not sure that the lambdas did what you think they did before. If you wanted CUDA lambdas you always had to add -expt-extentded-lambda AND -std=c++11 explicitly to the CMAKE_CXX_FLAGS. And no one was doing that in a standard build. So in the standard builds CUDA lambdas could not possibly have worked.

**Comment by crtrott** [comment]
So to some up what I am somewhat objecting to is that we make a setting the default which forces users to add certain flags to the CMAKE_CXX_FLAGS. You can't just do a KOKKOS_ENABLE_CUDA=ON and Trilinios_ENABLE_Tpetra=ON and get a working build.

**Comment by ibaned** [comment]
> So in the standard builds CUDA lambdas could not possibly have worked.

I've had a bit of experience with this, there are cases in which the lambdas would compile and run somewhat reasonably (e.g. `parallel_for` would work), but other things would break mysteriously (e.g. `parallel_reduce` with reduction values less than 64 bits on Pascal). Setting `Kokkos_ENABLE_Cuda_Lambda` fixed these things, but it was non-obvious that that was the solution. I think both before and after I already had `-expt-extended-lambda`. This may also have been with older Kokkos versions. Anyway its likely that Trilinos gets lucky and builds and passes some tests without all the proper settings, but it is definitely a dangerous way to work, so I'm in favor of adding as many checks as we can to ensure people have the flags they really need (which @mhoemmen has made progress on in Tpetra).

> You can't just do a KOKKOS_ENABLE_CUDA=ON and Trilinios_ENABLE_Tpetra=ON and get a working build.

CUDA builds of Trilinos were already much harder than that before lambdas came into the picture. You have to set `TPL_ENABLE_CUDA` and set `CMAKE_CXX_COMPILER` to `nvcc_wrapper` or do something with `OMPI_CXX` if thats how you want to combine the compilers. You still have to choose the right arch for the GPU you're targeting and pass that as a flag or modify `nvcc_wrapper` like I do.

So if I interpret what you're saying @crtrott , basically, we should still continue to treat lambdas as something not ready by default, keep having `#ifdef` functor alternatives to all our lambdas, instead of adding one more requirement for CUDA compilation which is the `-expt-extended-lambda` flag. Do we keep doing this until `nvcc` stops needing that flag?

**Comment by mhoemmen** [comment]
@crtrott If I make Tpetra check `CMAKE_CXX_FLAGS` for the lambda flag, and error out with a helpful message if it doesn't see the flag (possibly contingent on `CUDA_VERSION`), would that suffice?  Tpetra already checks for `Kokkos_ENABLE_Cuda_Lambda:BOOL=ON` in this way.

What I really would like to know, is whether lambdas will _always_ be a second-class feature that takes some special `--super-experimental-dont-use-but-here-you-go-daring-developer-you` flag, or whether we can assume that in the near future, they will be available by default.  If the former, then yes, I guess we have to work around if we care about CUDA.  If the latter, then I don't see the value of Trilinos implementing complicated work-arounds for not having lambdas.  When the giant computer full of GPUs shows up on our doorstep, CUDA will support lambdas natively, and everything will be good.

**Comment by ibaned** [comment]
@etphipp just made an interesting suggestion: https://github.com/trilinos/Trilinos/issues/1682#issuecomment-329588821 in the Trilinos issue tracking this same question. Basically, Trilinos/TriBITS should add `-expt-extended-lambda` to `CMAKE_CXX_FLAGS` if `Kokkos_ENABLE_Cuda_Lambda` is on.

**Comment by mhoemmen** [comment]
I would like to state for the record:
```
$ cd Trilinos/packages && grep -R KOKKOS_LAMBDA * | wc -l
233
$ cd kokkos && grep -R KOKKOS_LAMBDA * | wc -l
71
$ cd ../kokkos-kernels && grep -R KOKKOS_LAMBDA * | wc -l
37
$ cd ../tpetra && grep -R KOKKOS_LAMBDA * | wc -l
30
```
Tpetra has been conservative about `KOKKOS_LAMBDA`, compared to other packages.

**Comment by crtrott** [comment]
We will set up a meeting to discuss what to do with this in Trilinos.

**Comment by ibaned** [comment]
I've added a commit to this PR that does an additional check for whether `CMAKE_CXX_FLAGS` contains `-expt-extended-lambda`. This should make it totally safe to merge, the only behavior it changes is if all other flags except `Kokkos_ENABLE_Cuda_Lambda` are set correctly, then that one will be set to ON by default.
The latest commit also prints a message in all cases explaining the decision made and the reason for it.
We'd like to target merging this in the next promotion, Oct 6-12.

File: CMakeLists.txt [hunk 1]
```diff
@@ -33,6 +33,31 @@ TRIBITS_PACKAGE_DECL(Kokkos) # ENABLE_SHADOWING_WARNINGS)
...
+      set(KOKKOS_HAVE_CUDA_GEQ_75 TRUE)
+    endif()
+    if (CUDA_VERSION VERSION_GREATER_EQUAL "8.0")
+      set(KOKKOS_HAVE_CUDA_GEQ_80 TRUE)
+    endif()
+  endif()
+endif()
+
+set(Kokkos_ENABLE_Cuda_Lambda_DEFAULT OFF)
+if (KOKKOS_HAVE_CUDA_GEQ_75)
```
Comment (hcedwar) [inline CMakeLists.txt]: How does this pick up the `KOKKOS_HAVE_CUDA_GEQ_80` ?

File: CMakeLists.txt [hunk 2]
```diff
@@ -33,6 +33,31 @@ TRIBITS_PACKAGE_DECL(Kokkos) # ENABLE_SHADOWING_WARNINGS)
...
+  if (CMAKE_CXX_FLAGS MATCHES "-expt-extended-lambda")
```
Comment (hcedwar) [inline CMakeLists.txt]: Matches or contains?

Comment (ibaned) [reply CMakeLists.txt]: CMake doesn't have `CONTAINS`, `MATCHES` with a string as a regular expression essentially implements "contains". Also this has the nice property that `--expt-extended-lambda` should also be detected (two leading dashes).

File: CMakeLists.txt (hunk 1 above)
Comment (ibaned) [reply CMakeLists.txt]: see lines 36-45 above

Comment (hcedwar) [reply CMakeLists.txt]: ah yes, `VERSION_GREATER_EQUAL` not equal

File: CMakeLists.txt (hunk 2 above)
Comment (hcedwar) [reply CMakeLists.txt]: Oops, this option is likely version specific.  For example, CUDA 9.5 may not need the option.
Option detection should be tied to version or set/range of versions.

Comment (ibaned) [reply CMakeLists.txt]: well, right now its inside a check for version >= 7.5.
NVIDIA have not release a version yet that doesn't need this flag, so it shouldn't be too bad to wait until they do to add then "end" half of the range.

Comment (hcedwar) [reply CMakeLists.txt]: Flag is there for 7.5, 8.0, and 9.0.

**Pull request merged**